File: showcase/src/engine/splitter.ts
import {
  getArForms,
  getArNormForms,
  isArabic,
  lookup,
  normalizeAr,
} from "./lookupIndices";
import { FrequencyClass } from "./types";

const BASE_SEGMENT_COST = 1.0;
const UNKNOWN_PENALTY = 8.0;
const LENGTH_BONUS_PER_CHAR = -0.05;

const FREQ_BONUS: Record<FrequencyClass, number> = {
  common: -0.6,
  normal: -0.2,
  rare: 0.0,
};

function dpSegment(text: string): string[] {
  const arIndex = getArForms();
  const arNorm = getArNormForms();
  const n = text.length;

  const dpCost: number[] = new Array(n + 1).fill(Infinity);
  const dpPrev: number[] = new Array(n + 1).fill(-1);

  dpCost[0] = 0.0;
  dpPrev[0] = 0;

  for (let i = 1; i <= n; i++) {
    const startJ = Math.max(0, i - 30);
    for (let j = startJ; j < i; j++) {
      if (dpCost[j] === Infinity) continue;

      const substr = text.slice(j, i);
      if (substr.length < 2 && j > 0) continue;

      let entry = arIndex.get(substr);
      if (!entry) {
        entry = arNorm.get(normalizeAr(substr));
      }

      if (entry) {
        const bonus = FREQ_BONUS[entry.frequency as FrequencyClass] || 0.0;
        const cost =
          dpCost[j] + BASE_SEGMENT_COST + bonus + LENGTH_BONUS_PER_CHAR * substr.length;
        if (cost < dpCost[i]) {
          dpCost[i] = cost;
          dpPrev[i] = j;
        }
      } else {
        const cost = dpCost[j] + UNKNOWN_PENALTY + substr.length;
        if (cost < dpCost[i]) {
          dpCost[i] = cost;
          dpPrev[i] = j;
        }
      }
    }
  }

  if (dpCost[n] === Infinity) {
    return [text];
  }

  const segments: string[] = [];
  let pos = n;
  while (pos > 0) {
    const prev = dpPrev[pos];
    segments.push(text.slice(prev, pos));
    pos = prev;
  }

  segments.reverse();
  return segments;
}

export function split(fullName: string): string[] {
  if (!fullName || !fullName.trim()) return [];

  const text = fullName.trim();
  if (text.includes(" ")) {
    return text.split(/\s+/);
  }

  if (isArabic(text)) {
    const directEntry = lookup(text);
    if (directEntry) {
      return [text];
    }
    return dpSegment(text);
  }

  return [text];
}

File: java/egy-names/src/main/java/com/afify/egynames/engine/Splitter.java
package com.afify.egynames.engine;

import com.afify.egynames.index.LookupIndices;
import com.afify.egynames.model.Models;

import java.util.*;

public class Splitter {

    private static final double BASE_SEGMENT_COST = 1.0;
    private static final double UNKNOWN_PENALTY = 8.0;
    private static final double LENGTH_BONUS_PER_CHAR = -0.05;

    private static final Map<Models.FrequencyClass, Double> FREQ_BONUS = Map.of(
            Models.FrequencyClass.COMMON, -0.6,
            Models.FrequencyClass.NORMAL, -0.2,
            Models.FrequencyClass.RARE, 0.0
    );

    private static List<String> dpSegment(String text, String dataPath) {
        Map<String, Models.NameEntry> arIndex = LookupIndices.getArForms(dataPath);
        Map<String, Models.NameEntry> arNorm = LookupIndices.getArNormForms(dataPath);

        int n = text.length();
        double[] dpCost = new double[n + 1];
        int[] dpPrev = new int[n + 1];

        Arrays.fill(dpCost, Double.POSITIVE_INFINITY);
        Arrays.fill(dpPrev, -1);

        dpCost[0] = 0.0;
        dpPrev[0] = 0;

        for (int i = 1; i <= n; i++) {
            int startJ = i > 30 ? i - 30 : 0;
            for (int j = startJ; j < i; j++) {
                if (Double.isInfinite(dpCost[j])) continue;

                String substr = text.substring(j, i);
                if (substr.length() < 2 && j > 0) continue;

                Models.NameEntry entry = arIndex.get(substr);
                if (entry == null) {
                    entry = arNorm.get(LookupIndices.normalizeAr(substr));
                }

                if (entry != null) {
                    double bonus = FREQ_BONUS.getOrDefault(entry.frequency, 0.0);
                    double cost = dpCost[j] + BASE_SEGMENT_COST + bonus + LENGTH_BONUS_PER_CHAR * substr.length();
                    if (cost < dpCost[i]) {
                        dpCost[i] = cost;
                        dpPrev[i] = j;
                    }
                } else {
                    double cost = dpCost[j] + UNKNOWN_PENALTY + substr.length();
                    if (cost < dpCost[i]) {
                        dpCost[i] = cost;
                        dpPrev[i] = j;
                    }
                }
            }
        }

        if (Double.isInfinite(dpCost[n])) {
            return List.of(text);
        }

        List<String> segments = new ArrayList<>();
        int pos = n;
        while (pos > 0) {
            int prev = dpPrev[pos];
            segments.add(text.substring(prev, pos));
            pos = prev;
        }

        Collections.reverse(segments);
        return segments;
    }

    public static List<String> split(String fullName, String dataPath) {
        if (fullName == null || fullName.trim().isEmpty()) return List.of();

        String text = fullName.trim();
        if (text.contains(" ")) {
            return Arrays.asList(text.split("\\s+"));
        }

        if (LookupIndices.isArabic(text)) {
            Models.NameEntry entry = LookupIndices.lookup(text, dataPath);
            if (entry != null) {
                return List.of(text);
            }
            return dpSegment(text, dataPath);
        }

        return List.of(text);
    }
}
